src/NUnitEngine/nunit.engine/Services/TestSelectionParser.cs

```diff
@@ -0,0 +1,275 @@
+        private readonly Token OR_OP2 = new Token(TokenKind.Symbol, "||");
+        private readonly Token NOT_OP = new Token(TokenKind.Symbol, "!");
+
+        public string Parse(string input)
+        {
+            if (input == null || input == "")
+                return "";
+
+            _tokenizer = new Tokenizer(input);
+
+            return ParseFilterExpression();
+        }
+
+        /// <summary>
+        /// Parse a single term or an or expression, returning the xml
+        /// </summary>
+        /// <returns></returns>
+        public string ParseFilterExpression()
+        {
+            var terms = new List<string>();
+            terms.Add(ParseFilterTerm());
+
+            while (true)
+            {
+                Token op = Peek();
+                if (op != OR_OP1 && op != OR_OP2)
+                    break;
+
+                NextToken();
+                terms.Add(ParseFilterTerm());
+            }
+
+            if (terms.Count == 1)
+                return terms[0];
+
+            var sb = new StringBuilder("<or>");
+
+            foreach (string term in terms)
+                sb.Append(term);
+
+            sb.Append("</or>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single element or an and expression and return the xml
+        /// </summary>
+        public string ParseFilterTerm()
+        {
+            var elements = new List<string>();
+            elements.Add(ParseFilterElement());
+
+            while (true)
+            {
+                Token op = Peek();
+                if (op != AND_OP1 && op != AND_OP2)
+                    break;
+
+                NextToken();
+                elements.Add(ParseFilterElement());
+            }
+
+            if (elements.Count == 1)
+                return elements[0];
+
+            var sb = new StringBuilder("<and>");
+
+            foreach (string element in elements)
+                sb.Append(element);
+
+            sb.Append("</and>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single filter element such as a category expression
+        /// and return the xml representation of the filter.
+        /// </summary>
+        public string ParseFilterElement()
+        {
+            Token nextOp = Peek();
+            if (nextOp == LPAREN || nextOp == NOT_OP)
+                return ParseExpressionInParentheses();
+
+            string lhs = GetWord();
+
+            switch (lhs)
+            {
+                case "id":
+                    string op = GetOperator("=", "==");
+                    string values = GetIdList();
+                    return string.Format("<id>{0}</id>", values);
+
+                case "cat":
+                case "category":
+                case "method":
+                case "class":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
```

Comment:
No, they aren't. The GetOperator call says "Get me a token with TokenKind.Symbol that has one of the following Text values. What this allows me to do is to specify which operators are valid with which lhs name. They started out being different for each one but ended up all alike except for id, for which I only allow equality. However, there would be no harm in allowing the other operators for id as well, even though it wouldn't be very useful.

I'll do another pass through with this and the other comments in mind.